File: src/main/kotlin/me/fungames/jfortniteparse/ue4/assets/exports/mats/UUnrealMaterial.java
package me.fungames.jfortniteparse.ue4.assets.exports.mats;

import me.fungames.jfortniteparse.ue4.converters.CMaterialParams;

import java.util.List;

public interface UUnrealMaterial {
    void getParams(CMaterialParams params);

    default boolean isTextureCube() {
        return false;
    }

    default void appendReferencedTextures(List<UUnrealMaterial> outTextures, boolean onlyRendered) {
        CMaterialParams params = new CMaterialParams();
        getParams(params);
        params.appendAllTextures(outTextures);
    }

    String name();
}

File: src/main/kotlin/me/fungames/jfortniteparse/ue4/assets/exports/mats/UUnrealMaterial.kt
package me.fungames.jfortniteparse.ue4.assets.exports.mats

import me.fungames.jfortniteparse.ue4.assets.exports.UObject
import me.fungames.jfortniteparse.ue4.converters.CMaterialParams
import me.fungames.jfortniteparse.ue4.objects.uobject.FObjectExport

@ExperimentalUnsignedTypes
abstract class UUnrealMaterial : UObject {
    constructor() : super()
    constructor(exportObject: FObjectExport) : super(exportObject)

    abstract fun getParams(params: CMaterialParams)

    fun isTextureCube() = false

    open fun appendReferencedTextures(outTextures: MutableList<UUnrealMaterial>, onlyRendered: Boolean) {
        val params = CMaterialParams()
        getParams(params)
        params.appendAllTextures(outTextures)
    }
}